C-API: Add call stack queries (functions and mixins)

mgreter opened the pull request:
Same as https://github.com/sass/libsass/pull/2239 but rebased and adjusted after memory refactoring

mgreter commented:
Fixes https://github.com/sass/libsass/issues/1037 and https://github.com/sass/libsass/issues/1950

## Corresponding commits in [perl-libsass][1]:

- [import stack queries][2]
- [call stack queries (functions and mixins)][3]
- [sass variable environment access][4]

[1]: https://github.com/sass/perl-libsass/commits/latest
[2]: https://github.com/sass/perl-libsass/commit/00ec13fb1877268c22d091af01bd6249d87508d1
[3]: https://github.com/sass/perl-libsass/commit/196fbd310dc28f9b0b6aca105c0ac68e243f9149
[4]: https://github.com/sass/perl-libsass/commit/2a3fa42d00746a59c8cf8a708267d1b6c6b0229c

## Sample perl program

### test.pl
```pl
use strict;
use warnings;

use CSS::Sass qw(get_callee_stack_size);
use CSS::Sass qw(get_callee_type);
use CSS::Sass qw(get_callee_name);
use CSS::Sass qw(get_callee_path);
use CSS::Sass qw(get_callee_line);
use CSS::Sass qw(get_callee_column);
use CSS::Sass qw(get_callee_var);
use CSS::Sass qw(set_callee_var);

my $sass = CSS::Sass->new(
  sass_functions => {
    'get-callee-stack-size()' => sub { get_callee_stack_size() },
    'get-callee-type($back : 0)' => sub { get_callee_type($_[0]) },
    'get-callee-name($back : 0)' => sub { get_callee_name($_[0]) },
    'get-callee-path($back : 0)' => sub { get_callee_path($_[0]) },
    'get-callee-line($back : 0)' => sub { get_callee_line($_[0]) },
    'get-callee-column($back : 0)' => sub { get_callee_column($_[0]) },
    'get-callee-var($name, $back : 0)' => sub { get_callee_var($_[0], $_[1]) },
    'set-callee-var($name, $val, $back : 0)' => sub { set_callee_var($_[0], $_[1], $_[2]) },
    'get-callee($back: 0)' => sub {
      return sprintf("(%d) %s at %s (%d/%d)",
        get_callee_type($_[0]),
        get_callee_name($_[0]),
        get_callee_path($_[0]),
        get_callee_line($_[0]),
        get_callee_column($_[0]),
      );
    }
  }
);

print $sass->compile_file("test.scss");
```

### test.scss
```scss
@function baz($back: 0) {
  @return get-callee($back);
}

@function bar($back: 0) {
  @return baz($back);
}

@function foo($back: 0) {
  @return bar($back);
}

@mixin foo($back: 0) {
  @content;
}

A {
  $foo: "foo";
  @include foo {
    callee-5: foo(5);
    callee-4: foo(4);
    callee-3: foo(3);
    callee-2: foo(2);
    callee-1: foo(1);
    callee-0: foo(0);
    foo: get-callee-var("$foo");
    foo: set-callee-var("$foo", "new");
    foo: get-callee-var("$foo");
  }
}
```

### result
```css
A {
  callee-5: (0) foo at D:/.../perl-libsass/test.scss (19/12);
  callee-4: (0) @content at D:/.../perl-libsass/test.scss (14/3);
  callee-3: (1) foo at D:/.../perl-libsass/test.scss (21/15);
  callee-2: (1) bar at D:/.../perl-libsass/test.scss (10/11);
  callee-1: (1) baz at D:/.../perl-libsass/test.scss (6/11);
  callee-0: (2) get-callee at D:/.../perl-libsass/test.scss (2/11);
  foo: foo;
  foo: new; }
```

nschonni commented on src/context.cpp:
```diff
@@ -87,8 +87,9 @@ namespace Sass {
 
-    // add cwd to include paths
-    include_paths.push_back(CWD);
+    // Sass 3.4: The current working directory will no longer be placed onto the Sass load path by default.
```
Is it work wrapping in a complier directive for the SASS_LANG = 3.4?

mgreter replied:
Not sure. I was under the impression current master is going for 3.5. Side note: we could use a runtime test for 3.4 vs 3.5 behavior, but a compiler directive is not simple, since we have a string in that constant (string compare is not really supported by `#if` preprocessor directives).

mgreter replied:
Just realized that this change was done in ruby sass 3.4, not 3.5, therefore I think we should already have that removed. So to speak this is a bug in libsass 3.4. Maybe Or I did not get your comment?

nschonni replied:
You're right that this is 3.5 now, but I was thinking for ease of backporting changes to 3.4 while they are still both being maintained. I guess it would need to be a plain definition of `SASS_LANGUAGE_3_4`

mgreter replied:
I created a seperate commit for it, so it's better visible in this PR (for cherry-pick). IMO implementers are still free to add this on their own if the want to keep compatibility.

mgreter replied:
This is actually a bugfix as ruby sass 3.4 removed this behavior and we did not yet ...

xzyfer replied:
I understand, however it's a breaking change and 3.4 is stable. The ship has sailed on breaking changes in 3.4. Back ports are only happening for small Sass fixes.

nschonni commented on src/file.hpp:
```diff
@@ -113,7 +121,10 @@ namespace Sass {
   namespace File {
 
-    std::vector<Include> resolve_includes(const std::string& root, const std::string& file);
+    static std::vector<std::string> defaultExtensions = { ".scss", ".sass", ".css" };
```
Why include CSS files?

mgreter replied:
Same behavor [as before][1], also see https://github.com/sass/libsass/pull/754
[1]: https://github.com/sass/libsass/blob/master/src/file.cpp#L308-L310

nschonni replied:
You're right, but isn't it going away by 4 though? Might be another one to conditional out

mgreter replied:
IMO it should move into C-API to be configurable. This is just a small step in that direction.

nschonni commented on src/sass.cpp:
```diff
@@ -49,11 +66,52 @@ extern "C" {
+  char* ADDCALL sass_compiler_find_include (const char* file, struct Sass_Compiler* compiler)
+  {
+    // get the last import entry to get current base directory
+    // struct Sass_Options* options = sass_compiler_get_options(compiler);
```
Dead code?

mgreter replied:
Yep, commented due to unused var compiler warning.

nschonni commented on src/sass.cpp:
```diff
@@ -49,11 +66,52 @@ extern "C" {
+  char* ADDCALL sass_compiler_find_file (const char* file, struct Sass_Compiler* compiler)
+  {
+    // get the last import entry to get current base directory
+    // struct Sass_Options* options = sass_compiler_get_options(compiler);
```
Dead code?

mgreter replied:
Yep, commented due to unused var compiler warning.

mgreter commented:
Needs a rebase after https://github.com/sass/libsass/pull/2267 is merged!

mgreter commented:
## A few C-API remarks 

The C-API has grown quite a bit and we now expose quite a few arrays. Implementation for them is often very different. Since the whole C-API is behind function calls and pointers now, we can use C++ behind those functions. Therefore we could just use `std::vector` instead of our own primitive implementations (as i.e. for `include_paths`, which is really a linked list in order to grow dynamically). But due to where we came from, most things on the C-API are still just structs (which are allocated via calloc and deleted via free). In order to support C++ containers, those structs need to be upgraded to real classes and allocated via new and freed via delete (otherwise we leak). So this task would basically boil down to design and implement a C++ API for LibSass (which would be a nice addon).

## More CI tests (plugins)

I've added my libsass plugins to the CI process (Travis only so far). This allows most of the C-API to be tested via CI. I created a specific [libsass-tests][1] plugin to expose most C-API features as custom functions (work in progress). I was hoping to include a small sass-spec suite for each plugin, but unfortunately ran against https://github.com/sass/sass-spec/issues/1025. So for now each plugin contains just one spec test that is run and compared via diff in libsass Travis CI.

For reference, these are the tested plugins:

- https://github.com/mgreter/libsass-glob
- https://github.com/mgreter/libsass-math
- https://github.com/mgreter/libsass-digest*
- https://github.com/mgreter/libsass-tests

*digest needs this PR to compile/work!

[1]: https://github.com/mgreter/libsass-tests

typoworx-de commented:
How to use this in js using node-sass?
